lib/vsc/mympirun/common.py

```diff
@@ -0,0 +1,333 @@
+
+    taken from easybuild/tools/filetools.py, 6/7/2016
+    """
+    paths = os.environ.get('PATH', '').split(os.pathsep)
+    for path in paths:
+        cmd_path = os.path.join(path, cmd)
+        # only accept path is command is there, and both readable and executable
+        if os.access(cmd_path, os.R_OK | os.X_OK):
+            LOGGER.info("Command %s found at %s", cmd, cmd_path)
+            return cmd_path
+    LOGGER.warning("Could not find command '%s' (with permissions to read/execute it) in $PATH (%s)", cmd, paths)
+    return None
+
+
+def version_in_range(version, lower_limit, upper_limit):
+    """
+    Check whether version is in specified range
+
+    :param lower_limit: lower limit for version (inclusive), no lower limit if None
+    :param upper_limit: upper limit for version (exclusive), no upper limit if None
+    """
+    in_range = True
+    if lower_limit is not None and LooseVersion(version) < LooseVersion(lower_limit):
+        in_range = False
+    if upper_limit is not None and LooseVersion(version) >= LooseVersion(upper_limit):
+        in_range = False
+    return in_range
+
+
+class SchedKlass(object):
+
+    _sched_for = []  # classname is default added
+    _sched_environ_test = []
+    SCHED_ENVIRON_ID = None
+
+    HIDDEN = False
+
+    # factory methods for Sched. To add a new Sched class just create a new class that extends the cluster class
+    # see http://stackoverflow.com/questions/456672/class-factory-in-python
+    @classmethod
+    def _is_sched_for(cls, name=None):
+        """see if this class can provide support for sched class"""
+        if name is not None:
+            # add class name as default
+            return name in cls._sched_for + [cls.__name__]
+
+        # guess it from environment
+        totest = cls._sched_environ_test
+        if cls.SCHED_ENVIRON_ID is not None:
+            totest.append(cls.SCHED_ENVIRON_ID)
+
+        for envvar in totest:
+            envval = os.environ.get(envvar, None)
+            if not envval:
+                continue
+            else:
+                return True
+
+        return False
+
+
+class MpiKlass(object):
+
+    _mpirun_for = None
+    _mpiscriptname_for = []
+    _mpirun_version = None
+
+    HIDDEN = False
+    RUNTIMEOPTION = None
+
+    # factory methods for MPI
+    @classmethod
+    def _is_mpirun_for(cls, mpirun_path):
+        """
+        Check if this class provides support for active mpirun command.
+
+        @param cls: the class that calls this function
+        @return: True if mpirun is located in $EBROOT*, and if $EBVERSION* value matches version requirement
+        """
+        res = False
+
+        mpiname = cls._mpirun_for
+        if mpiname:
+            LOGGER.debug("Checking whether %s (MPI name: %s) matches %s", cls, mpiname, mpirun_path)
+
+            # first, check whether specified mpirun location is in $EBROOT<NAME>
+            mpiroot, mpiversion = eb_root_version(mpiname)
+            if mpiroot:
```

Comment:
Reverse the tests and bail if one thing needed is not present, to avoid a staircase of if's.

Reply:
Fail fast and return, rather than building some test tree.

Reply:
this is the original code (:duck:s)

Reply:
Still, fix it :) It'll be better in the long run.